### PR TITLE
Subclass TCN from Keras Layer

### PR DESCRIPTION
**Pull request by psomers3**
Similar to my pull request for the TF2.0 branch, I have redone it to work for pure keras as it was for TF 1. I have not been able to check this as much because the computer I work on for this does not have a gpu compatible for parallelization. So if you (or someone else) wants to vet it first, please do.

### CODE REVIEW

File: tcn/tcn.py [hunk 1]
```diff
@@ -121,30 +201,89 @@ def __init__(self,
...
+                                                          use_batch_norm=self.use_batch_norm,
+                                                          kernel_initializer=self.kernel_initializer,
+                                                          name='residual_block_{}'.format(len(self.residual_blocks))))
+                # build newest residual block
+                self.residual_blocks[-1].build(self.build_output_shape)
+                self.build_output_shape = self.residual_blocks[-1].res_output_shape
+
+        # this is done to force keras to add the layers in the list to self._layers
+        for layer in self.residual_blocks:
+            self.__setattr__(layer.name, layer)
```
Comment (beasteers) [inline tcn/tcn.py]: Are there any repercussions for having a slash in an attribute name? To my surprise, setattr with an invalid python identifier doesn't through an error so I guess it should be fine

File: tcn/tcn.py [hunk 2]
```diff
@@ -121,30 +201,89 @@ def __init__(self,
...
+        skip_connections = list()
+        for layer in self.residual_blocks:
+            x, skip_out = layer(x, training=training)
+            skip_connections.append(skip_out)
+
         if self.use_skip_connections:
-            x = keras.layers.add(skip_connections)
+            x = add(skip_connections)
         if not self.return_sequences:
             x = Lambda(lambda tt: tt[:, -1, :])(x)
```
Comment (beasteers) [inline tcn/tcn.py]: Use `self.lambda_layer`

File: tcn/tcn.py [hunk 3]
```diff
@@ -4,49 +4,128 @@
...
+
+    def call(self, inputs, training=None):
+        """
+
+        Returns: A tuple where the first element is the residual model tensor, and the second
+                 is the skip connection tensor.
+        """
+        x = inputs
+        for layer in self.residual_layers:
+            if type(layer) is SpatialDropout1D:
```
Comment (beasteers) [inline tcn/tcn.py]: use `isinstance(layer, SpatialDropout1D)` instead

File: tcn/tcn.py (hunk 1 above)
Comment (psomers3) [reply tcn/tcn.py]: I did not even think about that, but there doesn't seem to be any.

**Comment by philipperemy** [comment]
So I tested it and seems that it does not train. Probably because the weights are not registered in the TCN layer.

```
"""
#Trains a TCN on the IMDB sentiment classification task.
Output after 1 epochs on CPU: ~0.8611
Time per epoch on CPU (Core i7): ~64s.
Based on: https://github.com/keras-team/keras/blob/master/examples/imdb_bidirectional_lstm.py
"""
import numpy as np
from keras import Model, Input
from keras.datasets import imdb
from keras.layers import Dense, Dropout, Embedding
from keras.preprocessing import sequence

from tcn import TCN

max_features = 20000
# cut texts after this number of words
# (among top max_features most common words)
maxlen = 100
batch_size = 32

print('Loading data...')
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)
y_train = np.array(y_train)
y_test = np.array(y_test)

i = Input(shape=(maxlen,))
x = Embedding(max_features, 128)(i)
x = TCN(nb_filters=64,
        kernel_size=6,
        dilations=[1, 2, 4, 8, 16, 32, 64])(x)
x = Dropout(0.5)(x)
x = Dense(1, activation='sigmoid')(x)

model = Model(inputs=[i], outputs=[x])

model.summary()

# try using different optimizers and different optimizer configs
model.compile('adam', 'binary_crossentropy', metrics=['accuracy'])

print('Train...')
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=1,
          validation_data=[x_test, y_test])
```

**Comment by philipperemy** [comment]
Or even the one with the `Sequential` yields the same result:

```
"""
#Trains a TCN on the IMDB sentiment classification task.
Output after 1 epochs on CPU: ~0.8611
Time per epoch on CPU (Core i7): ~64s.
Based on: https://github.com/keras-team/keras/blob/master/examples/imdb_bidirectional_lstm.py
"""
import numpy as np
from keras import Sequential
from keras.datasets import imdb
from keras.layers import Dense, Dropout, Embedding
from keras.preprocessing import sequence

from tcn import TCN

max_features = 20000
# cut texts after this number of words
# (among top max_features most common words)
maxlen = 100
batch_size = 32

print('Loading data...')
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)
y_train = np.array(y_train)
y_test = np.array(y_test)

model = Sequential()
model.add(Embedding(max_features, 128, input_shape=(maxlen,)))
model.add(TCN(nb_filters=64,
              kernel_size=6,
              dilations=[1, 2, 4, 8, 16, 32, 64]))
model.add(Dropout(0.5))
model.add(Dense(1, activation='sigmoid'))

model.summary()

# try using different optimizers and different optimizer configs
model.compile('adam', 'binary_crossentropy', metrics=['accuracy'])

print('Train...')
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=1,
          validation_data=[x_test, y_test])
```

**Comment by philipperemy** [comment]
Done with:
Keras                2.3.1
tensorflow           1.13.2

**Comment by psomers3** [comment]
That is odd. I will have to check later. I made sure before that all the weights show up under trainable_weights at the highest level. And things trained for me. I will get back to you...

**Comment by philipperemy** [comment]
@psomers3 perfect! Thank you!

**Comment by psomers3** [comment]
Okay, so it looks like while I thought I was testing on tf1, I was not. But all is good now. Embarrassingly, the main problem was just that I was inheriting from the wrong “Layer”. I also found an error in my implementation of call, so its good that I needed to go over it again. 
Please give it another try.

**Comment by philipperemy** [comment]
@psomers3 awesome! Going to review it again :)

**Comment by philipperemy** [comment]
`25000/25000 [==============================] - 215s 9ms/step - loss: 0.8460 - accuracy: 0.6817 - val_loss: 0.3753 - val_accuracy: 0.8320`
Looks much better now! Will do my testing and will merge once it's done! Thanks again :)

**Comment by philipperemy** [comment]
Both tasks did not seem to converge well (I re-ran the script many times). I re-ran it 2 times on master with the same requirements (tensorflow gpu 1.14) and it worked all the time.
- https://github.com/philipperemy/keras-tcn/blob/master/tasks/adding_problem/main.py
- https://github.com/philipperemy/keras-tcn/blob/master/tasks/copy_memory/main.py


The PR:
```
_________________________________________________________________
Layer (type)                 Output Shape              Param #
=================================================================
input_1 (InputLayer)         (None, 600, 2)            0
_________________________________________________________________
tcn (TCN)                    (None, 24)                88248
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 25
_________________________________________________________________
activation_28 (Activation)   (None, 1)                 0
=================================================================
Total params: 88,273
Trainable params: 88,273
Non-trainable params: 0
_________________________________________________________________
```

Master
```
[.......................................]
lambda_1 (Lambda)               (None, 24)           0           add_10[0][0]
__________________________________________________________________________________________________
dense_1 (Dense)                 (None, 1)            25          lambda_1[0][0]
__________________________________________________________________________________________________
activation_28 (Activation)      (None, 1)            0           dense_1[0][0]
==================================================================================================
Total params: 88,273
Trainable params: 88,273
Non-trainable params: 0
__________________________________________________________________________________________________
```

=> They have the same number of parameters at least.

So maybe there's an error in the graph.

Master looks like this on adding task:
```
x_test[0:1] = [[[0.20054726 0.        ]
  [0.868526   0.        ]
  [0.39291642 0.        ]
  ...
  [0.70638923 0.        ]
  [0.470957   0.        ]
  [0.49922899 0.        ]]].
y_test[0:1] = [[0.48291087]].
2019-10-10 14:02:30.710002: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0
2019-10-10 14:02:30.810439: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
pred = [[-4.0173807]].
200000/200000 [==============================] - 173s 863us/step - loss: 1.6612 - val_loss: 0.0013
Epoch 2/500
x_test[0:1] = [[[0.20054726 0.        ]
  [0.868526   0.        ]
  [0.39291642 0.        ]
  ...
  [0.70638923 0.        ]
  [0.470957   0.        ]
  [0.49922899 0.        ]]].
y_test[0:1] = [[0.48291087]].
pred = [[0.45794693]].
200000/200000 [==============================] - 173s 863us/step - loss: 0.0014 - val_loss: 6.5209e-04
Epoch 3/500
x_test[0:1] = [[[0.20054726 0.        ]
  [0.868526   0.        ]
  [0.39291642 0.        ]
  ...
  [0.70638923 0.        ]
  [0.470957   0.        ]
  [0.49922899 0.        ]]].
y_test[0:1] = [[0.48291087]].
pred = [[0.48808497]].
200000/200000 [==============================] - 173s 863us/step - loss: 7.0014e-04 - val_loss: 4.4052e-04
```

The PR:
```
Train on 200000 samples, validate on 40000 samples
Epoch 1/500
x_test[0:1] = [[[0.85732818 0.        ]
  [0.03299431 0.        ]
  [0.75093291 0.        ]
  ...
  [0.50684793 0.        ]
  [0.34349418 0.        ]
  [0.22529049 0.        ]]].
y_test[0:1] = [[1.08891443]].
2019-10-10 14:02:55.708938: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0
2019-10-10 14:02:55.803069: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
pred = [[182.6935]].
200000/200000 [==============================] - 175s 875us/step - loss: 21782.0454 - val_loss: 0.1687
Epoch 2/500
x_test[0:1] = [[[0.85732818 0.        ]
  [0.03299431 0.        ]
  [0.75093291 0.        ]
  ...
  [0.50684793 0.        ]
  [0.34349418 0.        ]
  [0.22529049 0.        ]]].
y_test[0:1] = [[1.08891443]].
pred = [[0.8928762]].
200000/200000 [==============================] - 172s 859us/step - loss: 0.1684 - val_loss: 0.1675
```

Very huge loss at the beginning. Totally off and then fail to learn the task.

So I would guess there's an error in the graph.

Done with tensorflow-gpu 1.14.

**Comment by psomers3** [comment]
Well, I WAS looking for something to do today... so that works out. haha

**Comment by philipperemy** [comment]
:)

**Comment by psomers3** [comment]
whoops... I had passed the activation setting to the inner activations of the residual blocks. It becomes difficult to learn with a linear activation function there. I now get the same results as the current master you posted for the adding task.

**Comment by philipperemy** [comment]
Overall looks good to me! Doing some additional checks but I think we're close to the merge ;)

**Comment by philipperemy** [comment]
Ok let's merge this!
